backend/static/js/keywordSearch.js

```diff
@@ -0,0 +1,76 @@
+// this function removes the highlight
+function unHighlight(input_str){
+  return input_str.replaceAll(/\<span class\=\"keyword-search-highlight\"\>(.*?)\<\/span\>/gi, "$1");
+}
+
+// this function highlights the matching sections
+function highlightWords(input_str, words) {
+  let return_str = unHighlight(input_str.toLowerCase());
+  // find matching sections
+  let matching_sections = [];
+  for (word of words){
+    let startIndex = 0;
+    while ((index = return_str.indexOf(word, startIndex)) > -1){
+      matching_sections.push([index, index + word.length]);
+      startIndex = index + word.length;
+    }
+  }
+  if (matching_sections.length === 0){
+    return input_str;
+  }
+  // consolidate overlapping sections
+  matching_sections.sort((a,b) => a[0] - b[0]);
+  matching_sections = matching_sections.reduce((accu,elem) => {
+    const [last, ...rest] = accu;
+    if (elem[0] <= last[1]){
+      return [[last[0], Math.max(elem[1], last[1])], ...rest]
+    }
+    return [elem].concat(accu);
+  }, [matching_sections[0]]);
+  // replace matching sections with highlighting span
+  return_str = unHighlight(input_str);
+  matching_sections.forEach(([from, to]) => {
+    return_str = `${return_str.substring(0, from)}<span class="keyword-search-highlight">${return_str.substring(from,to)}</span>${return_str.substring(to)}`;
+  });
+  return return_str;
+}
+
+// this function gets triggered by the "oninput" event of the examples text input field id=example-keyword-search
+// it takes the input of this field and searches for the given keyword in the example queries
+// a example-query is a match if each keyword maches a substring of the example-name
+function filterExamples(event) {
+  const keywords = event.target.value
+    .toLowerCase()
+    .trim()
+    .split(' ')
+    .filter((keyword) => keyword !== '');
+  let hits = 0;
+  const exampleItems = $("ul#example-list .example-name").each(function(idx) {
+    const exampleText = $(this).text().trim();
+    if (keywords.every((keyword) => exampleText.toLowerCase().includes(keyword))){
+      $(this).parent().parent().show();
+      $(this).html(highlightWords(exampleText, keywords));
+      hits++;
+    }else {
+      $(this).parent().parent().hide();
+    }
+  });
+  if (hits === 0){
+    $("#empty-examples-excuse").show();
+  }
+  else {
+    $("#empty-examples-excuse").hide();
+  }
+}
+
+// this is taken from here: https://www.geeksforgeeks.org/implement-search-box-with-debounce-in-javascript/
+// its a decorator that makes sure a function is only called once within the given delay
```

Comment:
In which sense is this a "decorator"?

Reply:
With "decorator" i was referring to a design pattern of a "wrapper" function, a higher order function that consumes and returns a function.
This allows to add behavior *before* and *after* the consumed function.
In `python` these wrapper functions are called decorators.
Since the `debounce` funciton is a wrapper function that adds behaviour, without altering the behavior of the consumed function, i  thought its appropriate to call it a "decorator".

In the commit [enhancement: better docstrings for keyword search](https://github.com/ad-freiburg/qlever-ui/pull/83/commits/0175a3986c1699579ef29f3897724b8f90bf7163)  i changed the comment and dropped the mention of a decorator.